### ISSUE TITLE
Can't specify interface version more than once

### ISSUE THREAD

**Issue by EvilOatmeal** [mathquill/mathquill#464]
Imagine that you want to use two third-party scripts that uses MathQuill on the same page.

They both have to define the interface version they want to use, because they're third-party and don't know about the other scripts on the page.

But currently you can't specify the interface version more than once.

When the second third party script tries to specify the interface version the method to do so is not available.

```js
function first3rdParty(el, latex) {
	var mq = MathQuill.interfaceVersion(1);
	mq.MathField(el).latex(latex);
}

function second3rdParty(el, latex) {
	var mq = MathQuill.interfaceVersion(1);
	mq.MathField(el).latex(latex);
}
```

**Comment by laughinghan** [comment]
If they don't know about each other, wouldn't they load and .noConflict()
their own independent copies of MathQuill? That's how we'd deal with this
if instead of interface versioning we just had backwards-incompatible
versions of MathQuill, right?

Do you think it's important that `interfaceVersion()` return a new instance
of the MathQuill API object, with potentially a different interface version
from other instances? It's totally doable, it just complicates the API.

Han

On Thu, Sep 17, 2015 at 8:01 AM, Kenneth Sundqvist <notifications@github.com

**Comment by EvilOatmeal** [comment]
Do you mean like this?

```js
function first3rdParty(el, latex) {
	var mq = MathQuill.noConflict().interfaceVersion(1);
	mq.MathField(el).latex(latex);
}

function second3rdParty(el, latex) {
	var mq = MathQuill.noConflict().interfaceVersion(1);
	mq.MathField(el).latex(latex);
}
```

Because that doesn't work either. Same issue.

I would expect the global MQ to only be available to get interface versioned instances of MQ and not change itself. I should not have to worry about someone else or some other part of the code using it before I do.

If I'm not using third-party code but I can still only specify the version once I have to do it in some boot configuration, and I can't partially migrate my app to a new interface version easily.

Using "no conflict" before I request an interface version feels like "I have to undo things in case someone got there before I did so things don't get messed up".

Either way, if I'm right or wrong, the way it works now is not easy to use and will likely lead to errors that can only be solved by reading the documentation and using "workarounds by design" (no conflict). Keeping in mind that unlike jQuery you will be required to specify a version in MQ 1.0.0.

Interface versioning beyond the release version itself is a major complication of the API and cost to the users at the hypothetical gain of adoption rate of newer versions.

Sorry if I'm sounding overly negative about this, but I feel like so far you've gone half the way with it and what we have gained is confusion.

**Comment by EvilOatmeal** [comment]
I'll try to illustrate the inevitable complexity I'm talking about with a comparison to jQuery plugins.

Two jQuery plugins were created for jQuery versions with compatible APIs, so you can just add them both to the same page and they work.

Two MQ scripts use the same MQ InterVer, but since MQ requires you to specify the InterVer they both have to do so and they have to protect themselves from anyone else doing so.

jQuery plugins are "plug-and-pray" but MQ scripts require you to have intimate knowledge of all code or always guard yourself because you know of the potential issues that could occur.

The way that I mentioned with creating InterVer instances of MQ is also guarding yourself but it would be the only option available so you would not cause errors by not knowing you have to add these extra guards.

**Comment by laughinghan** [comment]
You don't sound overly negative at all, I'm glad you're engaging with me on this! I really want feedback on this since it's a bit unorthodox.

<blockquote>
Do you mean like this?
</blockquote>
<blockquote>
MQ scripts require you to have intimate knowledge of all code or always guard yourself because you know of the potential issues that could occur
</blockquote>
<blockquote>
cause errors by not knowing you have to add these extra guards
</blockquote>

Assuming that by "guard" you mean `.noConflict()`, I think you misunderstood (which is my bad! That means my idea is confusing!). I was thinking by analogy to two third-party scripts that depend on different versions of jQuery, which might be like:

```html
<script src="jquery-1.8.0.js"></script>
<script>
function first3rdParty() {
  var jQuery = jQuery.noConflict();
  // ...
}
</script>
<script src="jquery-1.9.0.js"></script>
<script>
function second3rdParty() {
  var jQuery = jQuery.noConflict();
  // ...
}
</script>
```

Right?

Analogously:

```html
<script src="mathquill.js"></script>
<script>
function first3rdParty() {
  var MathQuill = MathQuill.interfaceVersion(1).noConflict();
  // ...
}
</script>
<script src="mathquill.js"></script>
<script>
function second3rdParty() {
  var MathQuill = MathQuill.interfaceVersion(2).noConflict();
  // ...
}
</script>
```

That should work, right? But it's confusing, I take it?

--

> I would expect the global MQ to only be available to get interface versioned instances of MQ and not change itself. I should not have to worry about someone else or some other part of the code using it before I do.

Yeah, I'm not enthusiastic about mutating a global, so I wanted it as much as possible to be like a declaration. I even thought of letting you do like, `<script src="mathquill.js" data-mq-interver=1></script>`, but then I'd have to use `document.currentScript` or something and have a fallback and it just...seemed like trouble.

You're thinking about this differently. You're imagining `.interfaceVersion()` non-mutatively creates an instance of the MathQuill API of a given interface version, right? I don't hate that, I too hate mutation, I just didn't like that there'd be 2 similarly-named global variables for the API consumer to keep track of, assuming they do something like `window.MQ = MathQuill.interfaceVersion(2);`.

But you're telling me the mutative "declaration" is actually more confusing, not to mention multiple MathQuill scripts having to have their own copies of MathQuill is dumb (they might even step on each other's toes since there's data stored outside the DOM).

So, if we allow a single MathQuill script-include to be instantiated as different interface versions, how do instances of different interface versions interact? Like,

```js
var span = $('#math-field')[0];

var MQ1 = MathQuill.interfaceVersion(1);
var mathField1 = MQ1.MathField(span);

var MQ2 = MathQuill.interfaceVersion(2);
var mathField2 = MQ2(span);
```
Which interface version does `mathField2` expose? Should be 2, right? How do we implement that? Won't there be multiple API objects associated with one math field instance?

(This is already a problem, of course, if you do the thing I mention above where you have 2 script-includes for 2 different copies of MathQuill, they'll be mostly invisible to each other; but at least it would only come up if you did something as funky as script-including different copies of the same thing, rather than just calling some pure functions and using their results.)

> I feel like so far you've gone half the way with it and what we have gained is confusion.

Is the "full way" rather than "half way" that you're thinking of this non-mutatively-create-instance-with-given-interface-version way?

**Comment by EvilOatmeal** [comment]
If you don't return instances and instead mutate the global how can scripts use different InterVers of the same MQ global?

If I've understood it correctly the point of InterVer that is that MQ 1.x could support InterVer 1 and 2, and then MQ 2.x supports InterVer 2 and 3, for example. So you would only include MQ 1.x but could use both InterVer 1 and 2.

One issue that jQuery doesn't have but MQ does is that if you have control over all the code jQuery doesn't require you to use `noConflict` but MQ requires you to use `interfaceVersion`, so you have to declare it at least once, possibly in your boot code.

If you declare InterVer in the boot code it is likely visually disconnected from the code that then uses MQ, so it's harder to keep track of it when your code base is larger than what fits on the monitor.

If you don't declare it in the boot code, because you're third-party or like to keep things close, you have to protect yourself and maybe even others in some way.

> I even thought of letting you do like, `<script src="mathquill.js" data-mq-interver=1></script>`

I'm not a big fan of stuff like this because it's not commonly used so I have to think about it in a special way (thats also a part of why the whole InterVer thing is bugging me.)

This also makes it harder to work with different systems that handles the script includes for you, along the ways of WordPress and RequireJS.

> I just didn't like that there'd be 2 similarly-named global variables for the API consumer to keep track of, assuming they do something like `window.MQ = MathQuill.interfaceVersion(2);`.

That's true, but it is the choice of the consumer. The other option is to use `noConflict`, so you still have to keep track of things. The default option should be the safest option, and by default I mean if you hardly read the instructions and just plug it in and go with it if "it seems to work", because that's how the majority of programmers do things, unfortunately.

> how do instances of different interface versions interact

Either each InterVer declares how they are interoperable with other InterVers, or they don't assume any interoperability at all.

How do different things interact when they are used on the same thing? You pray that they do or they prevent you from using something else on the same thing once they got control of it. jQuery recommends that you namespace your events and use techniques like `$el.plugin("method")` to avoid conflicts, but there is still no guarantee.

My suggestion is to check the InterVer of the instance and if they're incompatible give a warning and not let the incompatible version use it.

> Is the "full way" rather than "half way" that you're thinking of this non-mutatively-create-instance-with-given-interface-version way?

Yes, well, the safer way I feel.

What would happen if someone else declared the InterVer before you and then you use MQ and assume a different InterVer is used?

What would the downsides be for the consumer if the global MQ would only let you retrieve instances or InterVer'd entry points?

Keeping track of that instance is one, but that is on the same level of protecting yourself from the wrong InterVer whenever you use MQ, in my opinion.

**Comment by EvilOatmeal** [comment]
Does InterVer block v0.10.0, or is it targeted for v1.0.0? It's a pretty big blocker to tackle.

**Comment by laughinghan** [comment]
It is big, but it's the only blocker for v0.10.0.

I want to make some minor breaking changes for v1.0.0 without breaking people who were on `dev`, that's how I came up with it in the first place.

**Comment by laughinghan** [comment]
> If you don't return instances and instead mutate the global how can scripts use different InterVers of the same MQ global?

You don't, just like you don't use different versions of the same jQuery global. To use different versions of jQuery, you have to load different copies.

> So you would only include MQ 1.x but could use both InterVer 1 and 2.

I was intending that you use only one of InterVer 1 or 2, never both.

> If you declare InterVer in the boot code it is likely visually disconnected from the code that then uses MQ, so it's harder to keep track of it when your code base is larger than what fits on the monitor.

Yes, I'm mildly worried about this; but which version of jQuery you're using is also visually disconnected from the code that uses jQuery, so I thought it could slide.

> > I even thought of letting you do like, `<script src="mathquill.js" data-mq-interver=1></script>`
>
> This also makes it harder to work with different systems that handles the script includes for you, along the ways of WordPress and RequireJS.

Indeed. Although, I dunno about WordPress but for RequireJS, the shim that'd currently be required to use MathQuill [lets you specify](http://requirejs.org/docs/api.html#config-shim) a function that's called at boot-time; and when MathQuill works properly with RequireJS we can use its [config info](http://requirejs.org/docs/api.html#config-moduleconfig) system.

> The default option should be the safest option

Of course, the point we're trying to figure out is what's the safest option.

<blockquote>
My suggestion is to check the InterVer of the instance and if they're incompatible give a warning and not let the incompatible version use it.
</blockquote>
<blockquote>
What would happen if someone else declared the InterVer before you and then you use MQ and assume a different InterVer is used?
</blockquote>

The point of InterVer is to have small incompatibilities between different interface versions. How about if API consumers could call `MathQuill.interfaceVersion()` multiple times, but it'd error if different versions were ever supplied? That's like your suggestion but errors as soon as possible.

> What would the downsides be for the consumer if the global MQ would only let you retrieve instances or InterVer'd entry points?
>
> Keeping track of that instance is one, but that is on the same level of protecting yourself from the wrong InterVer whenever you use MQ, in my opinion.

That's pretty much it, it made what I expect to be the common case more complex.

> ```js
> // NOTE: If we use MathQuill somewhere else we have to move the "interface version" declaration to some central place.
> ```

Is this hard? This is exactly what I was mildly worried about but figured couldn't be too hard, surely RequireJS can't be the only one with a config info system.

**Comment by EvilOatmeal** [comment]
> How about if API consumers could call MathQuill.interfaceVersion() multiple times, but it'd error if different versions were ever supplied?

I like this!

It simplifies authoring third party scripts a lot, and if you declare the InterVer yourself at boot and a third party declares a different InterVer you would know right away instead of getting stupid messages like `undefined is not a function`.

> Is [using a boot configuration] hard?

It's not hard at all, but it adds complexity. You have to know about it, and other people have to know about it too.

I think it's more common in RequireJS to do up front configuration, but in our system (concatenated Angular scripts) we only configure things like API keys at the Angular app boot, never library (interface) versions.

**Comment by laughinghan** [comment]
So I now think you were right in the first place. This `MathQuill.interfaceVersion()` declaration would require the interface version of MathQuill used in an app and its third-party dependencies to all be updated together in perfect lockstep, easily resulting in version dependency hell: if an app uses interface version 5 and it wants to use a new feature in an update of a dependency, but the updated dependency uses interface version 6, the app would have to update all its usages of MathQuill. If the app has other dependencies that use MathQuill, the app would have to update _all those other dependencies_ to use interface version 6, too. If they don't have updates to support 6, or the updates have changes the app doesn't want, the app may have to fork them, when they were otherwise working perfectly fine!

In such a situation a good third-party script would want to support a range of interface versions, probably by using a specific interface version and using a shim for the others in the range---which is exactly what MathQuill's interface versioning code is already doing internally!

Third-party plugins are exactly why jQuery can't do breaking changes, when being able to do breaking changes without breaking your code is exactly why I want to try interface versioning.

I went ahead and implemented `var MQ = MathQuill.getInterface(2)` semantics on #495, and though the changes were hundreds of lines, the net increase in source code was only 32 lines, which is pretty small right? There were 2 main sacrifices:
- There's no longer a unique API object per actual, "physical" instance of a math field or static math, instead `MQ()` (which was bound to interface version 2 by the above) creates a new API object whose interface version will be 2. This way a third-party script, or another module of your app, can do `var MQ = MathQuill.getInterface(1)` in its own scope, and its `MQ()` can create API objects whose interface version is 1, to manipulate the same math field you use interface version 2 to manipulate.
- Internally, registration of public API methods/classes (`MQ.MathField()`, `MQ.StaticMath()`) is more complicated now because `getInterface()` has to create them with a bound interface version and an isolated prototype chain (an API object whose interface version is 1 shouldn't be `instanceof` an `MQ()` whose interface version is 2).

The first breaking change using interface versioning was super easy, though, so that's a good sign: https://github.com/mathquill/mathquill/commit/2f78c8c5b5dd5ed7e4e12692759ffe0d1700e026#diff-8909a7456a5a30d96a556982356b4909

**Comment by EvilOatmeal** [comment]
Yes, that will solve my issues. Very nice, thank you!

**Comment by laughinghan** [comment]
Fixed by #495
